### PR TITLE
Fix Idefics2 padding when the first example has no image

### PR DESCRIPTION
**Pull request by Arnavsharma2**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48753) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48753)

Idefics2 raises `IndexError` for `[[], [image]]` because padding assumes the first batch row contains an image. Use the first available image for channels/device in both backends, preserving image outputs and zero masks for text-only examples.

Related to #31854

## Tests

3 failures -> 3 passes 

- `python -m pytest -q tests/models/idefics2/test_image_processing_idefics2.py`: 25 passed, 3 skipped.
- `python -m pytest -q tests/models/idefics2/test_processing_idefics2.py`: 31 passed, 31 skipped.
- `make fix-repo` and `make typing`: passed.

CPU only

## AI disclosure

AI helped in the development of tests and code

### CODE REVIEW

**Review by zucchini-nlp — COMMENTED** [review]
The support has been added already (in https://github.com/huggingface/transformers/pull/36682) and there is a test as well, can you double-check if the test is running on Idefics2 and provide a clear reproducer of the "bug"?

**Comment by zucchini-nlp** [comment]
If it is just a PR trying to solve an old GH issue, better to simply close it! The issue is too old and has been resolved

**Comment by Arnavsharma2** [comment]
> The support has been added already (in #36682) and there is a test as well, can you double-check if the test is running on Idefics2 and provide a clear reproducer of the "bug"?

Hi!

I double-checked, and [[], [image]] still raises IndexError in both Idefics2 image-processing backends.

I also ran test_processor_text_has_no_visual for Idefics2.

It is collected but skipped because the guard needs both images and videos in the processor signature.

Here's a small reproducer:

```
import numpy as np
from transformers import (
    Idefics2ImageProcessor,
    Idefics2ImageProcessorPil,
)

image = np.zeros((8, 10, 3), dtype=np.uint8)

for cls in (Idefics2ImageProcessor, Idefics2ImageProcessorPil):
    processor = cls(do_resize=False, do_image_splitting=False)
    try:
        result = processor([[], [image]], return_tensors="pt")
        print(cls.__name__, result.pixel_values.shape)
    except IndexError as error:
        print(cls.__name__, repr(error))
```

Both of them fail during padding when accessing the first image of the empty first row. My changes change it so that both return shape (2, 1, 3, 8, 10) with a zero pixel-attention mask for the text-only row.

**Comment by zucchini-nlp** [comment]
>  the guard needs both images and videos in the processor signature

nooooooo 😿  Oke, in that case would you be open to re-make the test and allow image-only inputs as well? That feature should be working in all VLMs, no matter if video is supported or not

**Comment by Arnavsharma2** [comment]
> > the guard needs both images and videos in the processor signature
> 
> nooooooo 😿 Oke, in that case would you be open to re-make the test and allow image-only inputs as well? That feature should be working in all VLMs, no matter if video is supported or not

Sure!

**Comment by Arnavsharma2** [comment]
> > the guard needs both images and videos in the processor signature
> 
> nooooooo 😿 Oke, in that case would you be open to re-make the test and allow image-only inputs as well? That feature should be working in all VLMs, no matter if video is supported or not

Just pushed the changes!

**Comment by zucchini-nlp** [comment]
We'll defi need to fix failing models now, there should be more than only Idefics2

**Comment by Arnavsharma2** [comment]
> We'll defi need to fix failing models now, there should be more than only Idefics2

Pushed the follow-up!
PI0 was also crashing on empty image rows, so I fixed that and added test coverage

NeoMME already supports mixed document batches, but inputs didn’t match its retrieval format. I modified its test to cover both direct inputs and its supported chat format

GLM-Image requires equal source-image counts across the batch for image generation. I kept that restriction and made a test that ensures it holds up that restriction!

Tests pass: 88 pass 103 skip

File: tests/models/idefics2/test_image_processing_idefics2.py [hunk 1]
```diff
@@ -319,6 +320,35 @@ def test_pixel_attention_mask(self):
...
+            np.arange(8 * 10 * 3, dtype=np.uint8).reshape(8, 10, 3),
+            np.arange(12 * 6 * 3, dtype=np.uint8).reshape(12, 6, 3),
+        ]
+        if input_type == "pil":
+            images = [Image.fromarray(image) for image in images]
+        elif input_type == "torch":
+            images = [torch.from_numpy(image).permute(2, 0, 1) for image in images]
+
+        for backend, image_processing_class in self.image_processing_classes.items():
+            for do_image_splitting in (False, True):
```
Comment (zucchini-nlp) [inline tests/models/idefics2/test_image_processing_idefics2.py]: not needed if we have a test 'has_no_visual'?

File: tests/models/pi0/test_processing_pi0.py [hunk 2]
```diff
@@ -95,6 +95,30 @@ def test_multi_camera_padding_and_masks(self):
         self.assertEqual(outputs["pixel_values"].shape[:2], torch.Size([2, 2]))
         self.assertTrue(torch.equal(outputs["pixel_attention_mask"], torch.tensor([[True, True], [True, False]])))
 
+    @require_torch
+    def test_padding_with_text_only_samples(self):
+        processor = self.get_processor()
+        processor.image_seq_length = 2
+        images = [[self.prepare_images_inputs(), self.prepare_images_inputs()], [self.prepare_images_inputs()]]
+        text = ["task a", "task b"]
```
Comment (zucchini-nlp) [inline tests/models/pi0/test_processing_pi0.py]: same q here

File: tests/test_processing_common.py [hunk 3]
```diff
@@ -789,38 +789,40 @@ def test_processor_text_has_no_visual(self):
...
+        inputs_dict_nested = {"images": image_inputs_nested}
+        modalities = ["image"]
 
         processing_kwargs = {"return_tensors": "pt", "padding": True}
         # Shouldn't sample when input is a decoded video without metadata (fpx/duration/etc.)
-        if "videos" in inputs_dict:
+        if "videos" in input_args and hasattr(processor, "video_processor"):
+            video_inputs = self.prepare_videos_inputs(batch_size=3)
+            inputs_dict_nested["videos"] = [[video] for video in video_inputs]
+            modalities.append("video")
```
Comment (zucchini-nlp) [inline tests/test_processing_common.py:816]: hmm oke, we dont have video-only models yet

**Review by zucchini-nlp — APPROVED** [review]
thanks! no need for new tests though right?

File: tests/models/idefics2/test_image_processing_idefics2.py (hunk 1 above)
Comment (Arnavsharma2) [reply tests/models/idefics2/test_image_processing_idefics2.py]: U r totally right, just removed it

File: tests/models/pi0/test_processing_pi0.py (hunk 2 above)
Comment (Arnavsharma2) [reply tests/models/pi0/test_processing_pi0.py]: Removed this as well

File: tests/test_processing_common.py (hunk 3 above)
Comment (Arnavsharma2) [reply tests/test_processing_common.py:816]: Yep! This covers image-only and image+video processors for now

**Review by zucchini-nlp — APPROVED** [review]
thx
